### ISSUE TITLE
Some C++20 deprecation warnings reported with -std=c++20 are not reported with -std=c++17 -Wc++20-compat or -Weverything

### ISSUE THREAD

**Issue by cpeterso** [llvm/llvm-project#55338]
Some C++20 deprecation warnings reported with `-std=c++20` are not reported with `-std=c++17 -Wc++20-compat` or `-Weverything`.

The following diagnostics are no-ops unless also compiled with -std=c++20:

-Wambiguous-reversed-operator
-Wdeprecated-pragma 
-Wdeprecated-this-capture
-Wdeprecated-volatile

This makes `-Wc++20-compat` not very useful.

### Steps to reproduce

Compare the output of compiling the following program with `clang++ -std=c++17 -Weverything` and then `clang++ -std=c++20 -Weverything`:

```
#include <atomic>

struct A {
  volatile int x;

  void test() {
    std::atomic<int> a = ATOMIC_VAR_INIT(1); // -Wdeprecated-pragma with -std=c++20

    [=]() {
      x; // -Wdeprecated-this-capture with -std=c++20 plus -Wdeprecated-this-capture or -Weverything
      x++; // -Wdeprecated-volatile with -std=c++20
    }();
  }

  bool operator==(const A& rhs) /* const */ { return x == rhs.x; } // -Wambiguous-reversed-operator with -std=c++20
};

int main() {
  A{} == A{};
  return 0;
}
```

### Expected behavior

I would expect that compiling the program with `clang++ -std=c++17 -Weverything` (or `clang++ -std=c++17 -Wc++20-compat`, `clang++ -std=c++17 -Wc++20-compat-pedantic`, or `clang++ -std=c++17 -Wambiguous-reversed-operator -Wdeprecated-pragma -Wdeprecated-this-capture -Wdeprecated-volatile`) would report the same warnings as compiling with `clang++ -std=c++20 -Weverything`.

### Actual behavior

Compiling the program with `clang++ -std=c++17 -Weverything` reports no warnings, but compiling with `clang++ -Weverything -std=c++20`will report four warnings:

```
test.cpp:7:26: warning: macro 'ATOMIC_VAR_INIT' has been marked as deprecated [-Wdeprecated-pragma]
    std::atomic<int> a = ATOMIC_VAR_INIT(1); // -Wdeprecated-pragma with -std=c++20
                         ^
/Users/chris/.mozbuild/clang/bin/../include/c++/v1/atomic:2709:43: note: macro marked 'deprecated' here
#  pragma clang deprecated(ATOMIC_VAR_INIT)
                                          ^
test.cpp:11:8: warning: increment of object of volatile-qualified type 'volatile int' is deprecated [-Wdeprecated-volatile]
      x++; // -Wdeprecated-volatile with -std=c++20
       ^
test.cpp:10:7: warning: implicit capture of 'this' with a capture default of '=' is deprecated [-Wdeprecated-this-capture]
      x; // -Wdeprecated-this-capture with -std=c++20 plus -Wdeprecated-this-capture or -Weverything
      ^
test.cpp:9:6: note: add an explicit capture of 'this' to capture '*this' by reference
    [=]() {
     ^
      , this
test.cpp:19:7: warning: ISO C++20 considers use of overloaded operator '==' (with operand types 'A' and 'A') to be ambiguous despite there being a unique best viable function [-Wambiguous-reversed-operator]
  A{} == A{};
  ~~~ ^  ~~~
test.cpp:15:8: note: ambiguity is between a regular call to this operator and a call with the argument order reversed
  bool operator==(const A& rhs) /* const */ { return x == rhs.x; } // -Wambiguous-reversed-operator with -std=c++20
       ^
```

**Comment by AaronBallman** [comment]
Picking one particular example, but this applies to all of them I believe: `ATOMIC_VAR_INIT` was deprecated in C++20, but it was not deprecated in C++17. So it's reasonable for us to diagnose in C++20 but not C++17. The `-std=c++17 -Wc++20-compat` case is interestingly different though -- that's asking for diagnostics about problems that will come up when switching to C++20. It's not entirely clear to me that deprecation warnings qualify as a problem that will come up. The code still works in C++20 the same way it worked in C++17, the only difference is a notification that the API is discouraged and may be removed in the future.

I don't think the lack of a deprecation warning in C++17 is a bug, and I (weakly, at the moment) don't think the behavior with the compat warnings enabled is a bug either.
